### PR TITLE
Corrected Xlib implementation

### PR DESCRIPTION
**Pull request by moses-palmer**
This pull request includes a few fixes from @jnv and @Talv and adds the following:

 *  Errors from *Xlib* are actually handled, and not just lost.
 *  All printable characters are supported without having to resort to symbol names, using the table found [here](https://www.cl.cam.ac.uk/~mgk25/ucs/keysyms.txt).

### CODE REVIEW

**Comment by pepijndevos** [comment]
Oh hey, what happened here, @moses-palmer? These looked like good changes. But it seems @SavinaRoja hasn't merged anything lately.

Hither says the repos is deleted, but I can still see the commits. Is there any way I can merge this?

**Comment by moses-palmer** [comment]
Ah, I actually removed the repo since the project seemed dead. Now it's up
again.

I started a similar project (pynput <https://github.com/moses-palmer/pynput>)
since the keyboard handling of *PyUserInput*, at least on *X*, appeared to
be architecturally somewhat lacking, and merging pull requests into
*PyUserInput* appeared to not happen.

2015-11-29 10:46 GMT+01:00 Pepijn de Vos <notifications@github.com>:

**Comment by pepijndevos** [comment]
I still can't merge it via GitHub. If you can create a new PR, I'll merge it right away. Or I'll take a look at your reps and do it from the command line later.

I'm happy to see people actively working on mouse and keyboard libraries for Python, but I'd be even more happy to see a concerted effort. Would it be possible to improve the architecture of PyUserInput without big breaking API changes?

I would be happy to give you commit access, but I don't think I have permission to do that on this repo. I think it would be healthy for PyUserInput to maybe become an Github organisation with a liberal stance on commit access.

What I'm seeing so far is that a lot of people have a need for this project for some sweet hack, contribute some fixes, and move on. I think neither me nor @SavinaRoja are currently actively using PyUserInput.

**Comment by moses-palmer** [comment]
I do agree that a concerted effort is better. However...

I started looking into libraries for this for my project virtual-touchpad
<https://github.com/moses-palmer/virtual-touchpad>, which currently uses an
in-repo implementation. Unfortunately the current master branch of
*PyUserInput* did not work for my use case, so I made the modifications in
the pull request. The changes worked initially, until I started testing it
with more complex use cases such as typing text using non-current keyboard
layouts. I also found the *API* slightly clunky.

My current attempt at a keyboard controller can be found here
<https://github.com/moses-palmer/pynput/tree/feature-keyabord-controller-wip>
for your reference.

To sum up: I'd actually rather complete *pynput* than work on *PyUserInput*.

2015-11-29 11:09 GMT+01:00 Pepijn de Vos <notifications@github.com>:

**Comment by pepijndevos** [comment]
Haha, this is pretty much exactly the project that started PyMouse. https://metapep.wordpress.com/2009/07/11/ipod-touch-wifi-mouse-with-python-web-app/ (this is not my current blog, it's that long ago)

Judging from the open issues, keyboard layout, internationalisation, non-printable characters, etc are a big problem, at least on some platforms. I did not write or use the keyboard part of this project, but if the API is broken, I would personally be open to a v2 with a better API and improved handling for international layouts.

But of course you're also free to take the code from this repo and make something better. It's open source after all. The mouse handling of this project is pretty much PyMouse with some patches and a different license.

**Comment by moses-palmer** [comment]
Heh, I guess that is what you need this functionality for these days :-)

The *APIs* are pretty similar---there's only that much to do with a mouse.
The code is based on the old code from *virtual-touchpad*, *PyUserInput* and
the its open issues.

I have joined the *PyUserInput* organisation on *github*, but I will also
continue working on *pynput*; I guess I enjoy writing libraries from
scratch.



2015-11-29 11:39 GMT+01:00 Pepijn de Vos <notifications@github.com>:
